examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,213 @@
+parser.add_argument('-n_particles_test', default=100, type=int)
+parser.add_argument('-n_batch', default=5000, type=int)
+parser.add_argument('-n_epoch', default=5000, type=int)
+parser.add_argument('-dtype', default='float32', type=str,
+                    choices=['float32', 'float64'])
+parser.add_argument('-dataset', default='boston_housing', type=str,
+                    choices=['boston_housing', 'protein_data'])
+
+
+class SVGP:
+
+    def __init__(self, hps, n_train, std_y_train):
+        self._hps = hps
+        self._kernel = RBFKernel(hps.n_covariates)
+        self._x_ph = x_ph = tf.placeholder(
+            hps.dtype, [None, hps.n_covariates], 'x')
+        self._y_ph = y_ph = tf.placeholder(
+            hps.dtype, [None], 'y')
+        self._z_pos = tf.get_variable(
+            'z/pos', [hps.n_z, hps.n_covariates], hps.dtype,
+            initializer=tf.random_uniform_initializer(-1, 1))
+        self._n_particles_ph = n_particles_ph = tf.placeholder(
+            tf.int32, [], 'n_particles')
+        n_batch = tf.cast(tf.shape(x_ph)[0], hps.dtype)
+
+        # TRAIN
+        # ELBO = E_q log (p(y|fx)p(fx|fz)p(fz) / p(fx|fz)q(fz))
+        # So we remove p(fx|fz) in both log_joint and latent
+        def log_joint(observed):
+            model, _ = SVGP.build_model(self, observed, n_particles_ph)
+            prior, log_p_y_given_fx = model.local_log_prob(['fZ', 'y'])
+            return prior + log_p_y_given_fx / n_batch * n_train
+
+        variational = self.build_variational(n_particles_ph)
+        [var_fZ, var_fX] = variational.query(
+            ['fZ', 'fX'], outputs=True, local_log_prob=True)
+        var_fX = (var_fX[0], tf.zeros_like(var_fX[1]))
+        lower_bound = zs.variational.elbo(
+            log_joint,
+            observed={'y': y_ph},
+            latent={'fZ': var_fZ, 'fX': var_fX},
+            axis=0)
+        self._elbo = lower_bound.sgvb()
+        optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
+        self._infer_op = optimizer.minimize(self._elbo)
+
+        # PREDICT
+        observed = {'fX': var_fX[0], 'y': y_ph}
+        model, y_inferred = SVGP.build_model(self, observed, n_particles_ph)
+        log_likelihood = model.local_log_prob('y')
+        std_y_train = tf.cast(std_y_train, hps.dtype)
+        self.log_likelihood = log_mean_exp(log_likelihood, 0) / n_batch - \
+            tf.log(std_y_train)
+        y_pred_mean = tf.reduce_mean(y_inferred.distribution.mean, axis=0)
+        self.pred_mse = tf.reduce_mean((y_pred_mean - y_ph) ** 2)
+        self.pred_mse *= std_y_train ** 2
+
+    @zs.reuse('model')
+    def build_model(self, observed, n_particles, full_cov=False):
+        '''
+        Build the SVGP model.
+        Note that for inference, we only need the diagonal part of Cov[Y], as
+        ELBO equals sum over individual observations.
+        For visualization etc we may want a full covariance. Thus the argument
+        `full_cov`.
```

Comment:
full_cov are passed to fx|fz, so must be true. So the argument is no longer needed.

Reply:
Actually it's usually false. It's still useful if we want to visualize the simultaneous credible band, instead of the concatenated pointwise credible band which is wider. Although this is indeed rare for high-dimensional input.